reactor-core/src/test/java/reactor/core/publisher/FluxFlatMapTest.java

```diff
@@ -46,6 +59,81 @@
+	@Ignore("causes too much GC pressure and OutOfMemoryError")
+	public void flatmapInnerShouldntRequestInFusionModeSync_gcHeavy() {
+		/*
+		See original reproduction case in https://github.com/reactor/reactor-core/issues/2318
+		This simplified test attempts to trigger the bad state of the inner subscription a bit more directly,
+		and allows asserting said state rather than detecting a consequence / symptom (NoSuchElementException).
+		As such it runs (a bit) faster but may reproduce a bit less reliably.
+		*/
+		final Scheduler scheduler = afterTest.autoDispose(Schedulers.newParallel("flatmapInnerShouldntRequestInFusionModeSync", 2));
+		final List<AssertQueueSubscription<Integer>> innerQueueSubscriptions = Collections.synchronizedList(new ArrayList<>());
+		final List<Integer> innerContent = Arrays.asList(1, 2, 3, 4, 5);
+
+		Flux.fromStream(IntStream.range(0, 32).boxed())
+		    .flatMap(number -> {
+			    final ExecutorService executor = Executors.newSingleThreadExecutor();
+			    return Flux.<Long>create(sink -> sink.onRequest(requested -> executor.submit(() -> LongStream.range(0, requested).forEach(sink::next))))
+					    .flatMap(x -> {
+						    AssertQueueSubscription<Integer> assertQueueSubscription = new AssertQueueSubscription<Integer>() {
+							    @Override
+							    public int requestFusion(int requestedMode) {
+								    if (requestedMode != Fuseable.NONE) {
+									    return Fuseable.SYNC;
+								    }
+								    return Fuseable.NONE;
+							    }
+						    };
+						    assertQueueSubscription.addAll(innerContent);
+
+						    innerQueueSubscriptions.add(assertQueueSubscription);
+						    return subscriber -> subscriber.onSubscribe(assertQueueSubscription);
+					    }, 8, 4)
+					    .publishOn(scheduler)
+					    .take(Duration.ofSeconds(5));
+		    })
+		    .blockLast();
+
+		//queuesubscription cannot be represented as string by assertj so we avoid allSatisfy, instead we manually loop
+		innerQueueSubscriptions.forEach(aqs -> assertThat(aqs.requested)
+				.withFailMessage("Inner publisher shouldn't be requested by FlatMapInner, got <request(%s)>", aqs.requested)
+				.isZero());
+	}
+
+	@Test
+	public void flatmapInnerShouldntRequestInFusionModeSync() {
+		/*
+		See original reproduction case in https://github.com/reactor/reactor-core/issues/2318
+		This simplified test directly triggers the fusion/request, reliably reproducing the
+		root cause behind the NoSuchElementError in an efficient way.
+		*/
+		final AssertQueueSubscription<Integer> inner = new AssertQueueSubscription<Integer>() {
+			@Override
+			public int requestFusion(int requestedMode) {
+				if ((requestedMode & Fuseable.SYNC) != 0) {
+					return Fuseable.SYNC;
+				}
+				return Fuseable.NONE;
+			}
+		};
+		inner.addAll(Arrays.asList(1, 2, 3, 4, 5));
+
+		FluxFlatMap.FlatMapMain<?, Integer> parent = new FluxFlatMap.FlatMapMain<>(
+				AssertSubscriber.create(0), v -> Mono.empty(), false, 8, Queues.small(), 4, Queues.small());
```

Comment:
nit: WDYT about doing "one arg per line" here?

Reply:
done. once approved I'll squash the last 3 commits to merge with rebase